process-engine/process_engine_meta
# ✨⚡️ Implement disposal pattern

**Changes:**

1. Implement repository disposal, by use of the `disposableDiscoveryTag` and the `IDisposable` interface.
2. Add some logging during initialization and disposal stages of each repository.

**Issues:**

Part of process-engine/process_engine_runtime#146

PR: #7

## How can others test the changes?

This is best viewed during the integrationtests. When the repository gets disposed through the bootstrapper, you'll see the corresponding log output.
It will look something like this:

![bildschirmfoto 2018-12-10 um 14 32 11](https://user-images.githubusercontent.com/15343316/49742462-e27ef600-fc98-11e8-8e5b-c4e654221fe7.png)

If multiple repositories try to dispose the same database connection, the `SequelizeConnectionManager` will recognize this and just dispose the connection once. All follow-up attempts to dispose the connection will just produce a message, informing the user that the connection was not found.

![bildschirmfoto 2018-12-11 um 08 40 08](https://user-images.githubusercontent.com/15343316/49785341-924b7680-fd20-11e8-8193-55271c5abd6b.png)


**Known Issue**
You will see this log output about three times during the early stages of the integrationtests:
![bildschirmfoto 2018-12-10 um 16 29 51](https://user-images.githubusercontent.com/15343316/49742436-d5fa9d80-fc98-11e8-8775-73fe00abe6e4.png)

This is because of some tests that use TimerBoundaryEvents. These are currently somewhat bugged, in that they do not interrupt the FlowNodeInstance they are attached to.
See here: process-engine/process_engine_runtime#176 

Since the test itself is finished, after the BoundaryEvent was triggered, the repositories will get disposed, even though the decorated FlowNode is still running.
As soon as that FlowNode finishes, it will attempt to persist its new state, which will fail, because the sequelize instance it attempts to use has already been disposed.

This bug is not related to the disposal pattern and should be regarded as a symptom of process-engine/process_engine_runtime#176 instead.

## PR-Checklist

Please check the boxes in this list after submitting your PR:


## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️      |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️      |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>
